theano/gpuarray/tests/check_dnn_conv.py

```diff
@@ -0,0 +1,1136 @@
+#!/usr/bin/env python
+
+# Without args, this script executes all its tests like `nosetests -vs`
+# python check_dnn_conv.py
+
+# If there is only one arg `infos`, this script prints some infos about
+# supported algorithms and data type configurations for current GPU and cuDNN version.
+# python check_dnn_conv.py infos
+
+# If there is only one arg `list`, this script prints all test cases without running them.
+# python check_dnn_conv.py list
+
+# Else, any arg will be directly passed to nosetests.
+# python check_dnn_conv.py -xvs  # nosetests: verbose mode, capture output, exit at first error.
+
+from __future__ import absolute_import, print_function, division
+
+import math
+import sys
+from itertools import product, chain
+
+import nose
+import numpy as np
+from nose.plugins.skip import SkipTest
+
+import theano
+import theano.tests.unittest_tools as utt
+from theano.compat import ifilter
+from theano.compile.ops import shape_i_op
+from theano.configdefaults import SUPPORTED_DNN_CONV_ALGO_RUNTIME
+from theano.gpuarray import cudnn_defs
+from theano.gpuarray.basic_ops import infer_context_name, as_gpuarray_variable, gpu_contiguous, GpuAllocEmpty
+from theano.gpuarray.dnn import GpuDnnConvDesc, GpuDnnConv, GpuDnnConvGradW, GpuDnnConvGradI, version, get_precision
+from theano.gpuarray.tests.config import mode_with_gpu, ref_cast
+from theano.tensor.nnet.abstract_conv import get_conv_output_shape, assert_conv_shape
+from theano.tensor.nnet.corr import CorrMM, CorrMM_gradInputs, CorrMM_gradWeights
+from theano.tensor.nnet.corr3d import Corr3dMM, Corr3dMM_gradInputs, Corr3dMM_gradWeights
+from theano.tensor.opt import Assert
+
+
+# We provide a special implementation of dnn_conv, dnn_gradweight and dnn_gradinput
+# that support alpha, beta and out as parameters.
```

Comment:
Do we want to move that to the main functions instead? It would be less test code and will help make this more maintained.

Reply:
That depends if we want to allow user to specify `alpha`, `beta` and `out`.

But these 3 parameters will be available only if user (or developer) directly calls cuDNN functions (dnn_conv, dnn_gradweight, dnn_gradinput, etc.). They are not currently available in CPU convolutions interface.

Reply:
Personally, I would do the change I propose and I would also do this on the CPU if we where continuing Theano. But to lower the maintenance, I would do it only on the GPU for now.

As this extend the user interface, I would prefer that we ask @abergeron or @lamblin comments on this.

Reply:
I don't think it is worth extending the interface for the abstract convolution and other implementations (nor adding `gemm` and `gemv` functions to provide explicit accumulation and scaling for `dot`).
If you think it would help maintenance, then we can either:
- do what you propose and expand `dnn_conv`, `dnn_gradweight` and `dnn_gradinput`, or
- move the definition with new names into `dnn.py`, and have the current function keep their interface, and call the new ones.

I would favour the second one, because:
- exposing the new options would create an expectation that all convolutions do the same
- exposing the new options means that we have to document them, have meaningful error checking and error messages, and tests that it behaves correctly in all cases (not only the ones in the test script)
- we do not have to worry about changing the interface and what the right order of parameters should be, for instance

Reply:
I was proposing to extend dnn_conv, ... with the new parameters at the end to don't break the interface.

I do not agree this would add the expectation that all convolution do the same. We had and still have some special cases under gpuarray.

But if you insist that this file isn't enough tests, then we should just keep this PR as is. I don't think the options you propose are worth the time of extra testing as we stop investing time in Theano.